test/e2e/tests/ratelimit.go

```diff
@@ -33,20 +33,48 @@ var RateLimitTest = suite.ConformanceTest{
 			gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
 			gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
 
-			// TODO: find a better to make sure ratelimit load new configuration
-			// just wait a bit more time for now
-			time.Sleep(3 * time.Second)
+			// should just send exactly 4 requests, and expect 429
+			firstThreeExpResp := http.ExpectedResponse{
+				Request: http.Request{
+					Path: "/",
+				},
+				Response: http.Response{
+					StatusCode: 200,
+				},
+				Namespace: ns,
+			}
+			firstThreeReq := http.MakeRequest(t, &firstThreeExpResp, gwAddr, "HTTP", "http")
+			if err := GotNTimesExpectedResponse(t, 3, suite.RoundTripper, firstThreeReq, firstThreeExpResp); err != nil {
+				t.Errorf("fail to get expected response at first three request: %v", err)
+			}
 
-			// TODO: should just send exactly 4 requests, and expect 429
-			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, http.ExpectedResponse{
+			lastFourthExpResp := http.ExpectedResponse{
 				Request: http.Request{
 					Path: "/",
 				},
 				Response: http.Response{
 					StatusCode: 429,
 				},
 				Namespace: ns,
-			})
+			}
+			lastFourthReq := http.MakeRequest(t, &lastFourthExpResp, gwAddr, "HTTP", "http")
+			if err := GotNTimesExpectedResponse(t, 1, suite.RoundTripper, lastFourthReq, lastFourthExpResp); err != nil {
+				t.Errorf("fail to get expected response at last fourth request: %v", err)
+			}
 		})
 	},
 }
+
+func GotNTimesExpectedResponse(t *testing.T, n int, r roundtripper.RoundTripper, req roundtripper.Request, resp http.ExpectedResponse) error {
```

Comment:
instead of doing an exact check of `n` requests, you might need to use some until/eventual  logic and wait for some timeout duration until all `n` requests ended up generating the expected response / status code.
because the e2e test might start sending traffic before the route was configured, etc

Reply:
thanks, right now i wait till get the first 200, and then fire the rest three requests

Reply:
thanks, re-read the logic and it looks good !